Support 5.9 kernel as on Fedora 33

The Linux kernel 5.9 is just released. Fedora 33 isn't released yet. And even 33 beta has kernel 5.8 by default.
But it's possible to install vanilla kernel 5.9 even onto Fedora 31 using this instruction: https://fedoraproject.org/wiki/Kernel_Vanilla_Repositories
What is the known problem?
The `make_request_fn` has been moved to struct `block_device_operations` and renamed as `submit_bio`
https://github.com/torvalds/linux/commit/c62b37d96b6eb3ec5ae4cbe00db107bf15aebc93
So even after the fix of #48 compilation of the module will fail like this:
```
[elastio@fedora31-amd64-build elastio-snap]$ make
make -C src
make[1]: входимо до каталогу «/home/elastio/elastio-snap/src»
if [ ! -f kernel-config.h ] || tail -1 kernel-config.h | grep -qv '#endif'; then mkdir configure-tests/feature-tests/build; ./genconfig.sh "5.9.1-36.vanilla.1.fc31.x86_64" "-w"; fi;
generating configurations for kernel-5.9.1-36.vanilla.1.fc31.x86_64
make[2]: входимо до каталогу «/home/elastio/elastio-snap/src/configure-tests/feature-tests»
make[3]: входимо до каталогу «/usr/src/kernels/5.9.1-36.vanilla.1.fc31.x86_64»
make[3]: Залишаю каталог "/usr/src/kernels/5.9.1-36.vanilla.1.fc31.x86_64"
make[2]: Залишаю каталог "/home/elastio/elastio-snap/src/configure-tests/feature-tests"
performing configure test: HAVE_BDOPS_OPEN_INT - not present
performing configure test: HAVE_BDOPS_OPEN_INODE - not present
performing configure test: HAVE_BDEV_STACK_LIMITS - not present
performing configure test: HAVE_BD_SUPER - present
performing configure test: HAVE_BIO_BI_REMAINING - not present
performing configure test: HAVE_BIO_BI_BDEV - not present
performing configure test: HAVE_BIO_BI_POOL - present
performing configure test: HAVE_BIO_ENDIO_1 - present
performing configure test: HAVE_BIO_ENDIO_INT - not present
performing configure test: HAVE_BIOSET_CREATE_3 - not present
performing configure test: HAVE_BIO_LIST - present
performing configure test: HAVE_BIOSET_INIT - present
performing configure test: HAVE_BIOSET_NEED_BVECS_FLAG - present
performing configure test: HAVE_BLK_ALLOC_QUEUE_MK_REQ_FN_NODE_ID - not present
performing configure test: HAVE_BLK_ALLOC_QUEUE_GFP_T - present
performing configure test: HAVE_BLKDEV_GET_BY_PATH - present
performing configure test: HAVE_BLKDEV_PUT_1 - not present
performing configure test: HAVE_BLK_SET_DEFAULT_LIMITS - present
performing configure test: HAVE_BLK_SET_STACKING_LIMITS - present
performing configure test: HAVE_BLK_STATUS_T - present
performing configure test: HAVE_BVEC_MERGE_DATA - not present
performing configure test: HAVE_BVEC_ITER - present
performing configure test: HAVE_COMPOUND_HEAD - present
performing configure test: HAVE___DENTRY_PATH - not present
performing configure test: HAVE_DENTRY_PATH_RAW - present
performing configure test: HAVE_D_UNLINKED - present
performing configure test: HAVE_ENUM_REQ_OP - not present
performing configure test: HAVE_ENUM_REQ_OPF - present
performing configure test: HAVE_FILE_INODE - present
performing configure test: HAVE_FMODE_T - present
performing configure test: HAVE_FOPS_FALLOCATE - present
performing configure test: HAVE_GENHD_FL_NO_PART_SCAN - present
performing configure test: HAVE_IOPS_FALLOCATE - not present
performing configure test: HAVE_INODE_LOCK - present
performing configure test: HAVE_KERNEL_READ_PPOS - present
performing configure test: HAVE_KERNEL_WRITE_PPOS - present
performing configure test: HAVE_MAKE_REQUEST_FN_INT - not present
performing configure test: HAVE_KERN_PATH - present
performing configure test: HAVE_MAKE_REQUEST_FN_VOID - not present
performing configure test: HAVE_MERGE_BVEC_FN - not present
performing configure test: HAVE_NOTIFY_CHANGE_2 - not present
performing configure test: HAVE_MNT_WANT_WRITE - present
performing configure test: HAVE_NOOP_LLSEEK - present
performing configure test: HAVE_PART_NR_SECTS_READ - not present
performing configure test: HAVE_PROC_CREATE_FN_FILE_OPERATIONS - not present
performing configure test: HAVE_PATH_PUT - present
performing configure test: HAVE_PROC_CREATE_FN_PROC_OPS - present
performing configure test: HAVE_SB_START_WRITE - present
performing configure test: HAVE_SUBMIT_BIO_WAIT - not present
performing configure test: HAVE_STRUCT_PATH - present
performing configure test: HAVE_SUBMIT_BIO_1 - present
performing configure test: HAVE_SYS_OLDUMOUNT - not present
performing configure test: HAVE_TASK_STRUCT_TASK_WORKS_HLIST - not present
performing configure test: HAVE_THAW_BDEV_INT - not present
performing configure test: HAVE_TASK_STRUCT_TASK_WORKS_CB_HEAD - present
performing configure test: HAVE_UAPI_MOUNT_H - present
performing configure test: HAVE_USER_PATH_AT - present
performing configure test: HAVE_UUID_H - present
performing configure test: HAVE_VFS_FALLOCATE - present
performing configure test: HAVE_VFS_UNLINK_2 - not present
performing configure test: HAVE_VZALLOC - present
make[2]: входимо до каталогу «/home/elastio/elastio-snap/src/configure-tests/feature-tests»
make[3]: входимо до каталогу «/usr/src/kernels/5.9.1-36.vanilla.1.fc31.x86_64»
make[3]: Залишаю каталог "/usr/src/kernels/5.9.1-36.vanilla.1.fc31.x86_64"
make[2]: Залишаю каталог "/home/elastio/elastio-snap/src/configure-tests/feature-tests"
performing sys_mount lookup
grep: /lib/modules/5.9.1-36.vanilla.1.fc31.x86_64/System.map: Permission denied
performing sys_umount lookup
grep: /lib/modules/5.9.1-36.vanilla.1.fc31.x86_64/System.map: Permission denied
performing sys_oldumount lookup
grep: /lib/modules/5.9.1-36.vanilla.1.fc31.x86_64/System.map: Permission denied
performing sys_call_table lookup
grep: /lib/modules/5.9.1-36.vanilla.1.fc31.x86_64/System.map: Permission denied
performing printk lookup
grep: /lib/modules/5.9.1-36.vanilla.1.fc31.x86_64/System.map: Permission denied
make -C /lib/modules/5.9.1-36.vanilla.1.fc31.x86_64/build M=/home/elastio/elastio-snap/src modules
make[2]: входимо до каталогу «/usr/src/kernels/5.9.1-36.vanilla.1.fc31.x86_64»
  CC [M]  /home/elastio/elastio-snap/src/elastio-snap.o
/home/elastio/elastio-snap/src/elastio-snap.c:546:41: помилка: unknown type name ‘make_request_fn’
  546 | static inline int elastio_snap_call_mrf(make_request_fn *fn, struct request_queue *q, struct bio *bio){
      |                                         ^~~~~~~~~~~~~~~
/home/elastio/elastio-snap/src/elastio-snap.c:848:2: помилка: unknown type name ‘make_request_fn’
  848 |  make_request_fn *sd_orig_mrf; //block device's original make request function
      |  ^~~~~~~~~~~~~~~
/home/elastio/elastio-snap/src/elastio-snap.c: У функції ‘snap_mrf_thread’:
/home/elastio/elastio-snap/src/elastio-snap.c:2749:9: помилка: неявне оголошення функції ‘elastio_snap_call_mrf’; ви мали на увазі ‘elastio_snap_get_mnt’? [-Werror=implicit-function-declaration]
 2749 |   ret = elastio_snap_call_mrf(dev->sd_orig_mrf, elastio_snap_bio_get_queue(bio), bio);
      |         ^~~~~~~~~~~~~~~~~~~~~
      |         elastio_snap_get_mnt
/home/elastio/elastio-snap/src/elastio-snap.c: У функції ‘tracing_mrf’:
/home/elastio/elastio-snap/src/elastio-snap.c:3091:2: помилка: unknown type name ‘make_request_fn’
 3091 |  make_request_fn *orig_mrf = NULL;
      |  ^~~~~~~~~~~~~~~
/home/elastio/elastio-snap/src/elastio-snap.c: На верхньому рівні:
/home/elastio/elastio-snap/src/elastio-snap.c:3177:53: помилка: unknown type name ‘make_request_fn’
 3177 | static int find_orig_mrf(struct block_device *bdev, make_request_fn **mrf){
      |                                                     ^~~~~~~~~~~~~~~
/home/elastio/elastio-snap/src/elastio-snap.c: У функції ‘__tracer_should_reset_mrf’:
/home/elastio/elastio-snap/src/elastio-snap.c:3210:6: помилка: ‘struct request_queue’ has no member named ‘make_request_fn’
 3210 |  if(q->make_request_fn != tracing_mrf) return 0;
      |      ^~
/home/elastio/elastio-snap/src/elastio-snap.c: На верхньому рівні:
/home/elastio/elastio-snap/src/elastio-snap.c:3224:92: помилка: unknown type name ‘make_request_fn’
 3224 | static int __tracer_transition_tracing(struct snap_device *dev, struct block_device *bdev, make_request_fn *new_mrf, struct snap_device **dev_ptr){
      |                                                                                            ^~~~~~~~~~~~~~~
/home/elastio/elastio-snap/src/elastio-snap.c: У функції ‘__tracer_setup_snap’:
/home/elastio/elastio-snap/src/elastio-snap.c:3607:2: помилка: неявне оголошення функції ‘blk_queue_make_request’; ви мали на увазі ‘blk_queue_max_segments’? [-Werror=implicit-function-declaration]
 3607 |  blk_queue_make_request(dev->sd_queue, snap_mrf);
      |  ^~~~~~~~~~~~~~~~~~~~~~
      |  blk_queue_max_segments
/home/elastio/elastio-snap/src/elastio-snap.c: У функції ‘__tracer_destroy_tracing’:
/home/elastio/elastio-snap/src/elastio-snap.c:3734:38: помилка: неявне оголошення функції ‘__tracer_transition_tracing’; ви мали на увазі ‘__tracer_destroy_tracing’? [-Werror=implicit-function-declaration]
 3734 |   if(__tracer_should_reset_mrf(dev)) __tracer_transition_tracing(NULL, dev->sd_base_dev, dev->sd_orig_mrf, &snap_devices[dev->sd_minor]);
      |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                      __tracer_destroy_tracing
/home/elastio/elastio-snap/src/elastio-snap.c: У функції ‘__tracer_setup_tracing’:
/home/elastio/elastio-snap/src/elastio-snap.c:3766:8: помилка: неявне оголошення функції ‘find_orig_mrf’ [-Werror=implicit-function-declaration]
 3766 |  ret = find_orig_mrf(dev->sd_base_dev, &dev->sd_orig_mrf);
      |        ^~~~~~~~~~~~~
cc1: деякі попередження вважаються помилками
make[3]: *** [scripts/Makefile.build:283: /home/elastio/elastio-snap/src/elastio-snap.o] Помилка 1
make[2]: *** [Makefile:1784: /home/elastio/elastio-snap/src] Помилка 2
make[2]: Залишаю каталог "/usr/src/kernels/5.9.1-36.vanilla.1.fc31.x86_64"
make[1]: *** [Makefile:14: default] Помилка 2
make[1]: Залишаю каталог "/home/elastio/elastio-snap/src"
make: *** [Makefile:24: driver] Помилка 2
```

--- a/.github/buildbox/Vagrantfile
+++ b/.github/buildbox/Vagrantfile
@@ -32,17 +32,19 @@ ENV_VARS_SCRIPT
 DISTROS = [ "debian8", "debian9", "debian10", "amazon2", "centos6", "centos7", "centos8", "fedora31", "fedora32" ]
 
 # NOTE: This environment variable should be set to enable triggers of the type "action".
-# The actions are used to make a lock around Vagrant::Action::Builtin::BoxAdd to avoid 
+# The actions are used to make a lock around Vagrant::Action::Builtin::BoxAdd to avoid
 # a problem, when 2 instances of the same box are "uping" and downloading new box (or version) image.
 # Vagrant uses same temporary file for all concurrent downloads. As result, it's corrupted.
-# Setting it just for the 'vagrnat up'. 
+# Setting it just for the 'vagrnat up'.
 ENV['VAGRANT_EXPERIMENTAL']='typed_triggers' if ARGV.include?('up')
 
 Vagrant.configure("2") do |config|
 
   DISTROS.each do |distro|
     box = "#{distro}-amd64-build"
-    config.vm.define "#{box}-#{ENV['RUNNER_NUM']}", autostart: false do |b|
+    instance = "#{box}-#{ENV['RUNNER_NUM']}"
+    next if ARGV.include?('up') && ARGV.none?("#{instance}")
+    config.vm.define "#{instance}", autostart: false do |b|
       b.vm.box = "#{box}"
       b.vm.box_url = [ "https://s3.eu-central-1.wasabisys.com/blobs-wasabi.elastio.dev/devboxes/master/#{box}.json" ]
       b.vm.box_version = $box_version

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -20,11 +20,9 @@ jobs:
           debian8, debian9, debian10,
           centos7, centos8,
           amazon2,
-          fedora31
-          # TODO: Bring back
-          # - Fedora 32 when kernel 5.9  is supported (issue https://github.com/elastio/elastio-snap/issues/57 is fixed)
+          fedora31, fedora32
           # - Fedora 33 when kernel 5.10 is supported.
-          # fedora32, fedora33
+          # fedora33
         ]
         arch: [ amd64 ]
 
@@ -49,6 +47,18 @@ jobs:
           AWS_DEFAULT_REGION: ${{ secrets.AWS_REGION }}
         run: .github/scripts/start_box.sh
 
+      - name: Boot Fedora 32 into kernel 5.9
+        if: "${{ matrix.distro == 'fedora32' }}"
+        run: |
+          vagrant ssh ${{env.INSTANCE_NAME}} -c '
+            sudo yum localinstall -y https://kojipkgs.fedoraproject.org//packages/kernel/5.9.16/100.fc32/x86_64/kernel-core-5.9.16-100.fc32.x86_64.rpm
+            sudo yum localinstall -y https://kojipkgs.fedoraproject.org//packages/kernel/5.9.16/100.fc32/x86_64/kernel-modules-5.9.16-100.fc32.x86_64.rpm
+            sudo yum localinstall -y https://kojipkgs.fedoraproject.org//packages/kernel/5.9.16/100.fc32/x86_64/kernel-5.9.16-100.fc32.x86_64.rpm
+            sudo yum localinstall -y https://kojipkgs.fedoraproject.org//packages/kernel/5.9.16/100.fc32/x86_64/kernel-devel-5.9.16-100.fc32.x86_64.rpm
+            sudo reboot now' || true
+          sleep 5
+        working-directory: ${{env.BOX_DIR}}
+
       - name: Build packages
         run: vagrant ssh ${{env.INSTANCE_NAME}} -c 'make ${PKG_TYPE}'
         working-directory: ${{env.BOX_DIR}}
@@ -58,7 +68,7 @@ jobs:
         working-directory: ${{env.BOX_DIR}}
 
       - name: Build kernel module
-        run: vagrant ssh ${{env.INSTANCE_NAME}} -c 'make'
+        run: vagrant ssh ${{env.INSTANCE_NAME}} -c 'sudo make'
         working-directory: ${{env.BOX_DIR}}
 
       - name: Install kernel module
@@ -79,7 +89,7 @@ jobs:
           cd $BOX_DIR && vagrant ssh ${{env.INSTANCE_NAME}} -c 'echo -e "n\np\n\n\n\nw" | sudo fdisk /dev/vdb'
 
       - name: Run tests (qcow2 disk)
-        run: vagrant ssh ${{env.INSTANCE_NAME}} -c "cd tests && sudo ./elio-test.sh /dev/vdb"
+        run: vagrant ssh ${{env.INSTANCE_NAME}} -c "cd tests && sudo ./elio-test.sh /dev/vdb1"
         working-directory: ${{env.BOX_DIR}}
         timeout-minutes: 5
 

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -21,7 +21,7 @@ jobs:
           centos7, centos8,
           amazon2,
           fedora31, fedora32
-          # - Fedora 33 when kernel 5.10 is supported.
+          # - Fedora 33 when kernel 5.11 is supported.
           # fedora33
         ]
         arch: [ amd64 ]

--- a/src/configure-tests/feature-tests/bdops_submit_bio.c
+++ b/src/configure-tests/feature-tests/bdops_submit_bio.c
@@ -1,0 +1,23 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+/*
+ * Copyright (C) 2020 Elastio Software Inc.
+ */
+
+// 5.9 <= kernel_version
+
+#include "includes.h"
+
+static blk_qc_t snap_submit_bio(struct bio *bio)
+{
+	return BLK_QC_T_NONE;
+}
+
+static inline void dummy(void){
+	struct bio b;
+	struct block_device_operations bdo = {
+		.submit_bio = snap_submit_bio,
+	};
+
+	bdo.submit_bio(&b);
+}

--- a/src/configure-tests/feature-tests/blk_mq_submit_bio.c
+++ b/src/configure-tests/feature-tests/blk_mq_submit_bio.c
@@ -1,0 +1,12 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+/*
+ * Copyright (C) 2020 Elastio Software Inc.
+ */
+
+#include "includes.h"
+
+static inline void dummy(void){
+	struct bio b;
+	blk_mq_submit_bio(&b);
+}

--- a/src/configure-tests/feature-tests/make_request_fn_in_queue.c
+++ b/src/configure-tests/feature-tests/make_request_fn_in_queue.c
@@ -1,0 +1,17 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+/*
+ * Copyright (C) 2020 Elastio Software Inc.
+ */
+
+// 5.8 <= kernel_version
+
+#include "includes.h"
+
+static inline void dummy(void){
+    make_request_fn *mk_rq_fn;
+	struct request_queue q = {
+		.make_request_fn = mk_rq_fn,
+	};
+	(void)q;
+}

--- a/src/configure-tests/feature-tests/thaw_bdev_int.c
+++ b/src/configure-tests/feature-tests/thaw_bdev_int.c
@@ -11,5 +11,5 @@ static inline void dummy(void){
 	struct block_device bd;
 	struct super_block sb;
 
-	if(thaw_bdev(&bd, &sb)) bd.bd_private = 0;
+	if(thaw_bdev(&bd, &sb)) bd.bd_disk = 0;
 }

--- a/src/configure-tests/symbol-tests
+++ b/src/configure-tests/symbol-tests
@@ -3,3 +3,4 @@ sys_umount
 sys_oldumount
 sys_call_table
 printk
+blk_mq_submit_bio

--- a/src/elastio-snap.c
+++ b/src/elastio-snap.c
@@ -37,7 +37,7 @@ MODULE_VERSION(ELASTIO_SNAP_VERSION);
 #include <uapi/linux/mount.h>
 #endif
 
-#ifdef HAVE_BLK_ALLOC_QUEUE_MK_REQ_FN_NODE_ID
+#if defined HAVE_BLK_ALLOC_QUEUE_MK_REQ_FN_NODE_ID || defined USE_BDOPS_SUBMIT_BIO
 #include <linux/blk-mq.h>
 #include <linux/percpu-refcount.h>
 #endif
@@ -530,34 +530,113 @@ static int elastio_snap_should_remove_suid(struct dentry *dentry)
 	#define vzalloc(size) __vmalloc(size, GFP_KERNEL | __GFP_HIGHMEM | __GFP_ZERO, PAGE_KERNEL)
 #endif
 
+#if !defined HAVE_MAKE_REQUEST_FN_IN_QUEUE && defined HAVE_BDOPS_SUBMIT_BIO
+	// Linux kernel version 5.9+
+	// make_request_fn has been moved from the request queue structure to the
+	// block_device_operations as submit_bio function.
+	// See https://github.com/torvalds/linux/commit/c62b37d96b6eb3ec5ae4cbe00db107bf15aebc93
+	#define USE_BDOPS_SUBMIT_BIO
+
+	// Prototype bdev->fops->submit_bio but with the name already used in the code
+	typedef blk_qc_t (make_request_fn) (struct bio *bio);
+#endif
+
+static inline unsigned long disable_page_protection(void);
+static inline void reenable_page_protection(unsigned long cr0);
+
+#ifndef USE_BDOPS_SUBMIT_BIO
+static inline make_request_fn* elastio_snap_get_bd_mrf(struct block_device *bdev){
+	return bdev->bd_disk->queue->make_request_fn;
+}
+
+static inline void elastio_snap_set_bd_mrf(struct block_device *bdev, make_request_fn *mrf){
+	bdev->bd_disk->queue->make_request_fn = mrf;
+}
+#else
+static inline make_request_fn* elastio_snap_get_bd_mrf(struct block_device *bdev){
+	return bdev->bd_disk->fops->submit_bio;
+}
+
+static inline void elastio_snap_set_bd_mrf(struct block_device *bdev, make_request_fn *mrf){
+	unsigned long cr0;
+
+	preempt_disable();
+	cr0 = disable_page_protection();
+	((struct block_device_operations *)bdev->bd_disk->fops)->submit_bio = mrf;
+	reenable_page_protection(cr0);
+	preempt_enable();
+}
+#endif
+
+static inline struct request_queue *elastio_snap_bio_get_queue(struct bio *bio);
+
 #ifdef HAVE_MAKE_REQUEST_FN_INT
 	#define MRF_RETURN_TYPE int
 	#define MRF_RETURN(ret) return ret
 
-static inline int elastio_snap_call_mrf(make_request_fn *fn, struct request_queue *q, struct bio *bio){
+static inline int __elastio_snap_call_mrf(make_request_fn *fn, struct request_queue *q, struct bio *bio){
 	return fn(q, bio);
 }
+
+static inline int elastio_snap_call_mrf(make_request_fn *fn, struct bio *bio){
+	return __elastio_snap_call_mrf(fn, elastio_snap_bio_get_queue(bio), bio);
+}
+
 #elif defined HAVE_MAKE_REQUEST_FN_VOID
 	#define MRF_RETURN_TYPE void
 	#define MRF_RETURN(ret) return
 
-static inline int elastio_snap_call_mrf(make_request_fn *fn, struct request_queue *q, struct bio *bio){
+static inline int __elastio_snap_call_mrf(make_request_fn *fn, struct request_queue *q, struct bio *bio){
 	fn(q, bio);
+	return 0;
+}
+
+static inline int elastio_snap_call_mrf(make_request_fn *fn, struct bio *bio){
+	__elastio_snap_call_mrf(fn, elastio_snap_bio_get_queue(bio), bio);
 	return 0;
 }
 #else
 	#define MRF_RETURN_TYPE blk_qc_t
 	#define MRF_RETURN(ret) return BLK_QC_T_NONE
 
+#ifndef USE_BDOPS_SUBMIT_BIO
+static inline int __elastio_snap_call_mrf(make_request_fn *fn, struct request_queue *q, struct bio *bio){
+	return fn(q, bio);
+}
+
+static inline int elastio_snap_call_mrf(make_request_fn *fn, struct bio *bio){
+	return __elastio_snap_call_mrf(fn, elastio_snap_bio_get_queue(bio), bio);
+}
+#endif
+
 #ifdef HAVE_BLK_ALLOC_QUEUE_MK_REQ_FN_NODE_ID
-static MRF_RETURN_TYPE elastio_snap_null_mrf(struct request_queue *q, struct bio *bio){
+// Linux version 5.8
+static inline MRF_RETURN_TYPE elastio_snap_null_mrf(struct request_queue *q, struct bio *bio){
 	percpu_ref_get(&q->q_usage_counter);
 	return blk_mq_make_request(q, bio);
 }
 #endif
+#endif
 
-static inline int elastio_snap_call_mrf(make_request_fn *fn, struct request_queue *q, struct bio *bio){
-	return fn(q, bio);
+#ifdef USE_BDOPS_SUBMIT_BIO
+// Linux version 5.9+
+
+#ifdef HAVE_BLK_MQ_SUBMIT_BIO
+// Use real blk_mq_submit_bio for kernels 5.9.0 - 5.9.1 where it was exported and add compat HAVE_BLK_MQ_SUBMIT_BIO
+static inline blk_qc_t elastio_blk_mq_submit_bio(struct bio *bio){
+	return blk_mq_submit_bio(struct bio *bio);
+}
+#else
+blk_qc_t (*elastio_blk_mq_submit_bio)(struct bio *) = (blk_qc_t (*)(struct bio *)) (BLK_MQ_SUBMIT_BIO_ADDR + (long long)(((void *)printk) - (void *)PRINTK_ADDR));
+#endif
+
+static inline MRF_RETURN_TYPE elastio_snap_null_mrf(struct bio *bio){
+	percpu_ref_get(&bio->bi_disk->queue->q_usage_counter);
+	return elastio_blk_mq_submit_bio(bio);
+}
+
+static inline int elastio_snap_call_mrf(make_request_fn *fn, struct bio *bio){
+	return fn(bio);
 }
 #endif
 
@@ -921,10 +1000,19 @@ static void elastio_snap_proc_stop(struct seq_file *m, void *v);
 static int elastio_snap_proc_open(struct inode *inode, struct file *filp);
 static int elastio_snap_proc_release(struct inode *inode, struct file *file);
 
+#ifdef USE_BDOPS_SUBMIT_BIO
+// Linux version 5.9+
+static MRF_RETURN_TYPE snap_mrf(struct bio *bio);
+#endif
+
 static const struct block_device_operations snap_ops = {
 	.owner = THIS_MODULE,
 	.open = snap_open,
 	.release = snap_release,
+#ifdef USE_BDOPS_SUBMIT_BIO
+// Linux version 5.9+
+	.submit_bio = snap_mrf,
+#endif
 };
 
 static const struct file_operations snap_control_fops = {
@@ -2775,7 +2863,7 @@ static int snap_mrf_thread(void *data){
 		//submit the original bio to the block IO layer
 		elastio_snap_bio_op_set_flag(bio, ELASTIO_SNAP_PASSTHROUGH);
 
-		ret = elastio_snap_call_mrf(dev->sd_orig_mrf, elastio_snap_bio_get_queue(bio), bio);
+		ret = elastio_snap_call_mrf(dev->sd_orig_mrf, bio);
 #ifdef HAVE_MAKE_REQUEST_FN_INT
 		if(ret) generic_make_request(bio);
 #endif
@@ -2989,7 +3077,7 @@ static int snap_trace_bio(struct snap_device *dev, struct bio *bio){
 	unsigned int bytes, pages, i = 0;
 
 	//if we don't need to cow this bio just call the real mrf normally
-	if(!bio_needs_cow(bio, dev->sd_cow_inode)) return elastio_snap_call_mrf(dev->sd_orig_mrf, elastio_snap_bio_get_queue(bio), bio);
+	if(!bio_needs_cow(bio, dev->sd_cow_inode)) return elastio_snap_call_mrf(dev->sd_orig_mrf, bio);
 
 	//the cow manager works in 4096 byte blocks, so read clones must also be 4096 byte aligned
 	start_sect = ROUND_DOWN(bio_sector(bio) - dev->sd_sect_off, SECTORS_PER_BLOCK) + dev->sd_sect_off;
@@ -3109,12 +3197,17 @@ out:
 	}
 
 	//call the original mrf
-	ret = elastio_snap_call_mrf(dev->sd_orig_mrf, elastio_snap_bio_get_queue(bio), bio);
+	ret = elastio_snap_call_mrf(dev->sd_orig_mrf, bio);
 
 	return ret;
 }
 
+#ifdef USE_BDOPS_SUBMIT_BIO
+// Linux version 5.9+
+static asmlinkage MRF_RETURN_TYPE tracing_mrf(struct bio *bio){
+#else
 static MRF_RETURN_TYPE tracing_mrf(struct request_queue *q, struct bio *bio){
+#endif
 	int i, ret = 0;
 	struct snap_device *dev;
 	make_request_fn *orig_mrf = NULL;
@@ -3139,15 +3232,37 @@ static MRF_RETURN_TYPE tracing_mrf(struct request_queue *q, struct bio *bio){
 	}
 
 call_orig:
-	if(orig_mrf) ret = elastio_snap_call_mrf(orig_mrf, q, bio);
+#ifdef USE_BDOPS_SUBMIT_BIO
+	// Linux version 5.9+
+	if(orig_mrf) ret = elastio_snap_call_mrf(orig_mrf, bio);
+	else if (bio->bi_disk->fops->submit_bio == tracing_mrf){
+		//original_mrf is not found, however bio's submit_bio is tracing_mrf. so, usual way is applicable
+		LOG_WARN("error finding original_mrf for the traced bio");
+		ret = elastio_snap_null_mrf(bio);
+	}else if(bio->bi_disk->fops->submit_bio){
+		LOG_WARN("error finding original_mrf, but bio's submit_bio is not empty");
+		ret = bio->bi_disk->fops->submit_bio(bio);
+	}else{
+		LOG_WARN("error finding original_mrf. all are empty");
+		ret = submit_bio_noacct(bio);
+	}
+#else
+	if(orig_mrf) ret = __elastio_snap_call_mrf(orig_mrf, q, bio);
 	else LOG_ERROR(-EFAULT, "error finding original_mrf");
+#endif
 out:
 	MRF_RETURN(ret);
 }
 
+#ifndef USE_BDOPS_SUBMIT_BIO
+// Linux version < 5.9
 static MRF_RETURN_TYPE snap_mrf(struct request_queue *q, struct bio *bio){
 	struct snap_device *dev = q->queuedata;
-
+#else
+// Linux version >= 5.9
+static MRF_RETURN_TYPE snap_mrf(struct bio *bio){
+	struct snap_device *dev = bio->bi_disk->queue->queuedata;
+#endif
 	//if a write request somehow gets sent in, discard it
 	if(bio_data_dir(bio)){
 		elastio_snap_bio_endio(bio, -EOPNOTSUPP);
@@ -3207,10 +3322,26 @@ static int find_orig_mrf(struct block_device *bdev, make_request_fn **mrf){
 	struct snap_device *dev;
 	struct request_queue *q = bdev_get_queue(bdev);
 
-	if(q->make_request_fn != tracing_mrf){
+	if(elastio_snap_get_bd_mrf(bdev) != tracing_mrf){
 #ifndef HAVE_BLK_ALLOC_QUEUE_MK_REQ_FN_NODE_ID
-		*mrf = q->make_request_fn;
+#ifdef USE_BDOPS_SUBMIT_BIO
+		// Linux version 5.9+
+		*mrf = elastio_snap_get_bd_mrf(bdev);
+		if (!*mrf){
+			if (elastio_blk_mq_submit_bio){
+				*mrf = elastio_snap_null_mrf;
+				LOG_DEBUG("original submit_bio is empty, set to elastio_snap_null_mrf");
+			}else{
+				LOG_ERROR(-EFAULT, "error finding original mrf, original submit_bio and elastio_snap_null_mrf both are empty");
+				return -EFAULT;
+			}
+		}
 #else
+		// Linux versions older than 5.8
+		*mrf = q->make_request_fn;
+#endif
+#else
+		// Linux version 5.8
 		if (q->make_request_fn) *mrf = q->make_request_fn;
 		else{
 			*mrf = elastio_snap_null_mrf;
@@ -3237,7 +3368,7 @@ static int __tracer_should_reset_mrf(const struct snap_device *dev){
 	struct snap_device *cur_dev;
 	struct request_queue *q = bdev_get_queue(dev->sd_base_dev);
 
-	if(q->make_request_fn != tracing_mrf) return 0;
+	if(elastio_snap_get_bd_mrf(dev->sd_base_dev) != tracing_mrf) return 0;
 	if(dev != snap_devices[dev->sd_minor]) return 0;
 
 	//return 0 if there is another device tracing the same queue as dev.
@@ -3295,13 +3426,14 @@ static int __tracer_transition_tracing(struct snap_device *dev, struct block_dev
 		LOG_DEBUG("starting tracing");
 		*dev_ptr = dev;
 		smp_wmb();
-		if(new_mrf) bdev->bd_disk->queue->make_request_fn = new_mrf;
+		if(new_mrf) elastio_snap_set_bd_mrf(bdev, new_mrf);
 	}else{
 		LOG_DEBUG("ending tracing");
-#ifdef HAVE_BLK_ALLOC_QUEUE_MK_REQ_FN_NODE_ID
-		if(new_mrf) bdev->bd_disk->queue->make_request_fn = new_mrf == elastio_snap_null_mrf ? NULL : new_mrf;
+#if defined HAVE_BLK_ALLOC_QUEUE_MK_REQ_FN_NODE_ID || defined USE_BDOPS_SUBMIT_BIO
+// Linux version 5.8
+		if(new_mrf) elastio_snap_set_bd_mrf(bdev, new_mrf == elastio_snap_null_mrf ? NULL : new_mrf);
 #else
-		if(new_mrf) bdev->bd_disk->queue->make_request_fn = new_mrf;
+		if(new_mrf) elastio_snap_set_bd_mrf(bdev, new_mrf);
 #endif
 		smp_wmb();
 		*dev_ptr = dev;
@@ -3654,9 +3786,17 @@ static int __tracer_setup_snap(struct snap_device *dev, unsigned int minor, stru
 	}
 
 #ifdef HAVE_BLK_ALLOC_QUEUE_GFP_T
-	//register request handler
 	LOG_DEBUG("setting up make request function");
+
+// For the Linux kernel version 5.9+:
+// The snap_mrf function is already set in the block_device_operations snap_ops struct
+// as submit_bio func. So, the request handler is already registered.
+// See a line below "dev->sd_gd->fops = &snap_ops;"
+#ifndef USE_BDOPS_SUBMIT_BIO
+	// Linux kernel version <= 5.6
+	// register request handler
 	blk_queue_make_request(dev->sd_queue, snap_mrf);
+#endif
 #endif
 
 	//give our request queue the same properties as the base device's
@@ -4956,13 +5096,28 @@ static asmlinkage long oldumount_hook(char __user *name){
 #define X86_CR0_WP (1UL << 16)
 #endif
 
-static inline void disable_page_protection(unsigned long *cr0) {
-	*cr0 = read_cr0();
-	write_cr0(*cr0 & ~X86_CR0_WP);
+static inline void wp_cr0(unsigned long cr0) {
+#ifdef USE_BDOPS_SUBMIT_BIO
+	// Enabling/disabling approach with usage of the write_cr0 function stopped to work somewhere starting from the kernels 5.X (maybe 5.3)
+	// after this patch https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=8dbec27a242cd3e2816eeb98d3237b9f57cf6232
+	// Hence there is a workaround.
+	// The USE_BDOPS_SUBMIT_BIO have to be replaced with something else if this workaround is necessary for some other 5.X kernels.
+	// Now it's used just for the 5.9+ kernels.
+	__asm__ __volatile__ ("mov %0, %%cr0": "+r" (cr0));
+#else
+	write_cr0(cr0);
+#endif
 }
 
-static inline void reenable_page_protection(unsigned long *cr0) {
-	write_cr0(*cr0);
+static inline unsigned long disable_page_protection() {
+	unsigned long cr0;
+	cr0 = read_cr0();
+	wp_cr0(cr0 & ~X86_CR0_WP);
+	return cr0;
+}
+
+static inline void reenable_page_protection(unsigned long cr0) {
+	wp_cr0(cr0);
 }
 
 static void **find_sys_call_table(void){
@@ -4998,13 +5153,13 @@ static void restore_system_call_table(void){
 		LOG_DEBUG("restoring system call table");
 		//break back into the syscall table and replace the hooks we stole
 		preempt_disable();
-		disable_page_protection(&cr0);
+		cr0 = disable_page_protection();
 		restore_syscall(__NR_mount, orig_mount);
 		restore_syscall(__NR_umount2, orig_umount);
 #ifdef HAVE_SYS_OLDUMOUNT
 		restore_syscall(__NR_umount, orig_oldumount);
 #endif
-		reenable_page_protection(&cr0);
+		reenable_page_protection(cr0);
 		preempt_enable();
 	}
 }
@@ -5022,13 +5177,13 @@ static int hook_system_call_table(void){
 
 	//break into the syscall table and steal the hooks we need
 	preempt_disable();
-	disable_page_protection(&cr0);
+	cr0 = disable_page_protection();
 	set_syscall(__NR_mount, orig_mount, mount_hook);
 	set_syscall(__NR_umount2, orig_umount, umount_hook);
 #ifdef HAVE_SYS_OLDUMOUNT
 	set_syscall(__NR_umount, orig_oldumount, oldumount_hook);
 #endif
-	reenable_page_protection(&cr0);
+	reenable_page_protection(cr0);
 	preempt_enable();
 	return 0;
 }

